envisage/extension_point.py

```diff
@@ -264,3 +282,278 @@ def _get_extension_registry(self, obj):
+        ----------
+        obj : HasTraits
+            The object on which an ExtensionPoint is defined.
+        trait_name : str
+            The name of the trait for which ExtensionPoint is defined.
+        """
+        cache_name = self._get_cache_name(trait_name)
+        old = obj.__dict__.get(cache_name, Undefined)
+        new = (
+            _ExtensionPointValue(
+                _get_extensions(obj, trait_name),
+                object=obj,
+                name=trait_name,
+            )
+        )
+        obj.__dict__[cache_name] = new
+        obj.trait_property_changed(trait_name, old, new)
+
+
+class _ExtensionPointValue(TraitList):
+    """ _ExtensionPointValue is the list being returned while retrieving the
+    attribute value for an ExtensionPoint trait.
+
+    This list returned for an ExtensionPoint acts as a proxy to query
+    extensions in an ExtensionRegistry for a given extension point id. Users of
+    ExtensionPoint expect to handle a list-like object, and expect to be able
+    to listen to "mutation" on the list. The ExtensionRegistry remains to be
+    the source of truth as to what extensions are available for a given
+    extension point ID.
+
+    Users are not expected to mutate the list directly. All mutations to
+    extensions are expected to go through the extension registry to maintain
+    consistency. With that, all methods for mutating the list are nullified,
+    unless it is used internally.
+
+    The requirement to support ``observe("name:items")`` means this list,
+    associated with `name`, cannot be a property that gets recomputed on every
+    access (enthought/traits/#624), it needs to be cached. As with any
+    cached quantity, it needs to be synchronized with the ExtensionRegistry.
+
+    The assumption on the internal values being synchronized with the registry
+    breaks down if the extension registry is mutated before listeners
+    are hooked up between the extension point and the registry. This sequence
+    of events is difficult to enforce. Therefore we always resort to the
+    extension registry for querying values.
+
+    Parameters
+    ----------
+    iterable : iterable
+        Iterable providing the items for the list
+    obj : HasTraits
+        The object on which an ExtensionPoint is defined.
+    trait_name : str
+        The name of the trait for which ExtensionPoint is defined.
+    """
+
+    def __init__(self, iterable=(), *, object, name):
+        """ Reimplemented TraitList.__init__
+
+        Parameters
+        ----------
+        object : HasTraits
+            The object on which an ExtensionPoint is defined.
+        trait_name : str
+            The name of the trait for which ExtensionPoint is defined.
+        """
+        super().__init__(iterable)
+
+        # Flag to control access for mutating the list. Only internal
+        # code can mutate the list. See _sync_values
+        self._internal_use = False
+
+        self._object = object
```

Comment:
I was wondering if this has to be a weakref... and I am afraid it does to avoid cycle reference, and I will need to implement the pickling logic for the weakref.

Reply:
I am going to remove this: It only serves to support out-of-sync scenario when the listeners are not connected. This is probably not a scenario we need to worry about (?), I will note this as a limitation as the effort and maintenance burden to keep these back references are too much to be worth it.